internal/tiltfile2/tiltfile_state.go

```diff
@@ -0,0 +1,292 @@
+	// build functions
+	dockerBuildN = "docker_build"
+	fastBuildN   = "fast_build"
+
+	// k8s functions
+	k8sYamlN     = "k8s_yaml"
+	k8sResourceN = "k8s_resource"
+	portForwardN = "port_forward"
+
+	// file functions
+	localGitRepoN = "local_git_repo"
+	localN        = "local"
+	readFileN     = "read_file"
+	kustomizeN    = "kustomize"
+)
+
+func (s *tiltfileState) builtins() skylark.StringDict {
+	r := make(skylark.StringDict)
+	add := func(name string, fn func(thread *skylark.Thread, fn *skylark.Builtin, args skylark.Tuple, kwargs []skylark.Tuple) (skylark.Value, error)) {
+		r[name] = skylark.NewBuiltin(name, fn)
+	}
+
+	add(dockerBuildN, s.dockerBuild)
+	add(fastBuildN, s.fastBuild)
+
+	add(k8sYamlN, s.k8sYaml)
+	add(k8sResourceN, s.k8sResource)
+	add(portForwardN, s.portForward)
+
+	add(localGitRepoN, s.localGitRepo)
+	add(localN, s.local)
+	add(readFileN, s.skylarkReadFile)
+	add(kustomizeN, s.kustomize)
+	return r
+}
+
+const unresourcedName = "k8s_yaml"
+
+func (s *tiltfileState) assemble() ([]*k8sResource, error) {
+	hadUnresourced := len(s.k8sUnresourced) > 0
+
+	images, err := s.findUnresourcedImages()
+	if err != nil {
+		return nil, err
+	}
+	for _, image := range images {
+		if _, ok := s.imagesByName[image.Name()]; !ok {
+			// only expand for images we know how to build
+			continue
+		}
+		target, err := s.findExpandTarget(image)
+		if err != nil {
+			return nil, err
+		}
+		if err := s.extractImage(target, image); err != nil {
+			return nil, err
+		}
+	}
+
+	if hadUnresourced {
+		r, err := s.makeK8sResource(unresourcedName)
+		if err != nil {
+			return nil, err
+		}
+		r.k8s = s.k8sUnresourced
+	}
+
+	assembledImages := map[string]bool{}
+	for _, r := range s.k8s {
+		if err := s.validateK8s(r, assembledImages); err != nil {
+			return nil, err
+		}
+	}
+
+	for k, _ := range s.imagesByName {
+		if !assembledImages[k] {
+			return nil, fmt.Errorf("image %v is not used in any resource", k)
+		}
+	}
+
+	return s.k8s, nil
+}
+
+func (s *tiltfileState) validateK8s(r *k8sResource, assembledImages map[string]bool) error {
+	var images []reference.Named
+	for _, e := range r.k8s {
+		entityImages, err := e.FindImages()
+		if err != nil {
+			return err
+		}
+		images = append(images, entityImages...)
+	}
+
+	for _, image := range images {
+		if _, ok := s.imagesByName[image.Name()]; !ok {
+			continue
+		}
+		if r.imageRef == "" {
+			r.imageRef = image.Name()
+		} else if r.imageRef == image.Name() {
+			continue
+		} else {
+			return fmt.Errorf("resource %q contains two images to be built: %q, %q. You can use k8s_yaml to include a lot of yaml and then Tilt will create resources automatically. If this doesn't solve your case (e.g. you have one pod that has two images that Tilt updates), please reach out so we can understand and prioritize", r.name, r.imageRef, image.Name())
+		}
+	}
+
+	if len(r.k8s) == 0 && r.name != unresourcedName {
+		if r.imageRef == "" {
+			return fmt.Errorf("resource %q: no matching resource", r.name)
```

Comment:
huh? "resource" gets confusing here. do you mean that "we expected resource X to have an associated docker ref, but none found"?